### PR TITLE
Introduce RuntimeHelpers.GetObjectHeader and Optimize Object.GetHashCode() with it

### PR DESCRIPTION
**Pull request by EgorBo**
A GC-hole free version of https://github.com/dotnet/runtime/pull/55273, should slightly speed up object.GetHashCode, e.g.:

Example:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
int Test(Prog o) => o.GetHashCode();
```

Current codegen:
```asm
; Method Prog:Test(Prog):int:this (FullOpts)
       sub      rsp, 40
       cmp      byte  ptr [rdx], dl
       mov      rcx, rdx
       call     System.Runtime.CompilerServices.RuntimeHelpers:GetHashCode(System.Object):int
       nop      
       add      rsp, 40
       ret      
; Total bytes of code: 20
```

New codegen:
```asm
; Method Prog:Test(Prog):int:this (FullOpts)
       sub      rsp, 40
       cmp      byte  ptr [rdx], dl
       mov      eax, dword ptr [rcx-0x04] ;; <---
       mov      edx, eax
       and      edx, 0xC000000
       cmp      edx, 0xC000000
       jne      SHORT G_M29166_IG04
       and      eax, 0x3FFFFFF
       jmp      SHORT G_M8578_IG05
G_M8578_IG04:  ;; offset=0x0020
       mov      rcx, rdx
       call     System.Runtime.CompilerServices.RuntimeHelpers:InternalGetHashCode(System.Object):int
G_M8578_IG05:  ;; offset=0x0028
       nop      
       add      rsp, 40
       ret      
; Total bytes of code: 46
```

This should allow us to convert Monitor.Enter/Exit to C# as well (thin lock, etc).

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
I tried non-intrinsic version ealier and measured a significant regression.

**Comment by EgorBo** [comment]
> I tried non-intrinsic version ealier and measured a significant regression.

with pinning? no wonder then 🙂

**Comment by huoyaoyuan** [comment]
> with pinning? no wonder then 🙂

No pinning. I was not sure about how GC deals with exterior pointer. I reverted it then in #97641

File: src/coreclr/classlibnative/bcltype/objectnative.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ NOINLINE static INT32 GetHashCodeHelper(OBJECTREF objRef)
 
 // Note that we obtain a sync block index without actually building a sync block.
 // That's because a lot of objects are hashed, without requiring support for
-FCIMPL1(INT32, ObjectNative::GetHashCode, Object* obj) {
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/classlibnative/bcltype/objectnative.cpp:40]: Leaving that up-for-grabs, wasn't the point of this PR

**Comment by EgorBo** [comment]
> > with pinning? no wonder then 🙂
> 
> No pinning. I was not sure about how GC deals with exterior pointer. I reverted it then in #97641

I don't see why my version could be slower (it's not according to benchmarks) - the codegen has a small CQ problem (https://github.com/dotnet/runtime/issues/101000) but I am not even sure it can change the result.

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 2]
```diff
@@ -125,8 +143,34 @@ public static unsafe void PrepareMethod(RuntimeMethodHandle method, RuntimeTypeH
         [MethodImpl(MethodImplOptions.InternalCall)]
         public static extern void PrepareDelegate(Delegate d);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetHashCode(object? o)
```
Comment (AustinWise) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:147]: You could apply a similar treatment to `TryGetHashCode` below.

**Comment by huoyaoyuan** [comment]
> I don't see why my version could be slower

Yeah, I realized it may need intrinsic handling like retrieving method table.

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:147]: Thanks! Applied

**Comment by EgorBo** [comment]
Looks like this PR is a regression when object has a syncblock and I don't think I can surface `PassiveGetSyncBlock` into C#

**Comment by jkotas** [comment]
> This should allow us to convert Monitor.Enter/Exit to C# as well (thin lock, etc).

We would be need both reads and interlocked writes for this.

To enable rewriting methods like there in C#, I think it would be better to look into general pinning optimizations (#63397) rather than specialized intrinsics.

**Comment by jkotas** [comment]
> surface PassiveGetSyncBlock into C#

It can be done once we have a good reason to.

**Pull request closed without merging**